### PR TITLE
tests that run on master pull requests

### PR DESCRIPTION
**Pull request by gn5**
tests:
- build singularity images
- run pipeline and compare output with upstream fork (Connor lab)
- run Sanger profile and check it ran `bwa` with 4 cpus and `ivar` without `-e`

### CODE REVIEW

File: .github/workflows/pull_request.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    - name: Build pipeline Singularity containers
+      run: |
+       echo $(which singularity)
+       bash scripts/build_singularity_containers.sh
+    - name: install Conda
+      run: bash .github/scripts/install_conda.sh
+    - name: install Nextflow via Conda
+      run: |
+       export PATH=/opt/conda/bin:$PATH
+       conda install -c bioconda nextflow
```
Comment (sb10) [inline .github/workflows/pull_request.yml:24]: Why the use of bioconda to install nextflow? Why not a script to install java 8 instead, then `curl -s https://get.nextflow.io | bash`?

File: .github/scripts/run_tests.sh [hunk 2]
```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
+set -eo pipefail
+export PATH=/opt/conda/bin:$PATH
+
+# write log as github Action artifact
+mkdir artifacts
+echo run tests >> artifacts/test_artifact.log
+
+# run current pull request code
```
Comment (sb10) [inline .github/scripts/run_tests.sh]: Change to "run in illumina mode with defaults"

File: .github/scripts/run_tests.sh [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+singularity --version
+NXF_VER=20.03.0-edge nextflow run ./main.nf \
+       -profile singularity \
+       --directory $PWD/.github/data/ \
+       --illumina \
+       --prefix test
+cp .nextflow.log artifacts/
+
+# run upstream connor-lab fork
```
Comment (sb10) [inline .github/scripts/run_tests.sh]: This is going to end up as a PR in connor-lab, so this test doesn't make sense. Once connor-lab is tagged with releases, perhaps the test could compare to the previous release. For now you could checkout commit 42a79999f11304671b9d22e4959b0167b2130944 to compare against.

File: .github/scripts/run_tests.sh [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+else
+  echo no differences found between pull request and upstream fork       
+fi
+
+# test --sanger profile
+mv results results_singularity_profile
+rm -rf work && rm -rf .nextflow*
+# there are only 2 available cpus in the github runner execution
+sed -i s'/cpus = 4/cpus = 2/'g conf/coguk/sanger.config
+NXF_VER=20.03.0-edge nextflow run ./main.nf \
```
Comment (sb10) [inline .github/scripts/run_tests.sh]: Does the test fail if this exits non-zero?

File: .github/scripts/run_tests.sh [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	|| exit 1
+# check that sanger profile activated params.allowNoprimer = false:
+find work -name .command.sh \
+    | xargs cat | grep 'ivar trim -e' \
+    && exit 1 \
+	|| echo "sanger profile: did NOT ran ivar trim -e" 
+
+# test conda profile
+# install Conda environments of the pipeline
+#source .github/scripts/install_conda_envs.sh
```
Comment (sb10) [inline .github/scripts/run_tests.sh]: This shouldn't be done as part of this test, when you uncomment the rest of the test

File: .github/scripts/install_conda_envs.sh [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+set -eo pipefail
+
+conda env create -f environment-illumina.yml
```
Comment (sb10) [inline .github/scripts/install_conda_envs.sh]: I don't think this file is needed?

File: .github/scripts/install_go.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+set -eo pipefail
+
+# install Go compiler
```
Comment (sb10) [inline .github/scripts/install_go.sh]: What is go needed for? Is this file needed?

File: .github/scripts/install_go.sh [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+sudo rm -r /usr/local/go*
+
+# install Singularity dependencies
```
Comment (sb10) [inline .github/scripts/install_go.sh]: This is duplicated from install_singularity.sh

File: .github/scripts/install_singularity.sh [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    uuid-dev \
+    libgpgme11-dev \
+    squashfs-tools \
+    libseccomp-dev \
+    wget \
+    pkg-config \
+    git \
+    cryptsetup
+
+echo $(which go)
```
Comment (sb10) [inline .github/scripts/install_singularity.sh:18]: Are go-related lines needed in this file?

**Review by sb10 — CHANGES_REQUESTED** [review]
Additional tests:
Check that `--bed bef.file --ref bwa-indexed-ref.fa` workflow completes, has same output as without these options, but doesn't run the articDownloadScheme process (no `git clone` command is run), and doesn't run `bwa index`. The test can copy the bed file, ref file and index files from the results directory of the first test.

Having gotten the conda profile tests to work, confirm the same results using the `--cache` option, and that the given cache dir has contents, and there is no conda cache dir in the work dir.

Test `--cram` mode works, and gives same output as with the fastq inputs.

**Comment by sb10** [comment]
Thanks for the PR. Once the above review is resolved, not quite sure how we test. Suggestions welcome, but how about:
1. I merge the PR to wtsi-team112 master, but you don't delete your master_test branch.
2. I create some PRs that break the workflow in various ways to make sure the tests catch the errors.
3. If the tests need improving, I force-push wtsi-team112 master back to being even with upstream (so losing your test commits), then you make the improvements on your master_test branch and do a new PR. Then 1-3 are repeated until it's finished.

File: .github/workflows/pull_request.yml (hunk 1 above)
Comment (gn5) [reply .github/workflows/pull_request.yml:24]: since the test of the Conda cache requires Conda, I figured I could use it to install Nextflow as well with a one-liner. Do you prefer I install java 8/NF with a script?

File: .github/scripts/run_tests.sh (hunk 2 above)
Comment (gn5) [reply .github/scripts/run_tests.sh]: done

File: .github/scripts/run_tests.sh (hunk 3 above)
Comment (gn5) [reply .github/scripts/run_tests.sh]: ok, done.

File: .github/scripts/run_tests.sh (hunk 4 above)
Comment (gn5) [reply .github/scripts/run_tests.sh]: yes the GitHub workflow would be tagged as failed (I added `set -eo pipefail` in all test scripts.)

File: .github/scripts/run_tests.sh (hunk 5 above)
Comment (gn5) [reply .github/scripts/run_tests.sh]: done: I have split the tests in separate GitHub workflow tasks.

File: .github/scripts/install_go.sh (hunk 7 above)
Comment (gn5) [reply .github/scripts/install_go.sh]: I remove that script. Go is needed to install singularity but it is already installed in the GitHub runners.

File: .github/scripts/install_singularity.sh (hunk 9 above)
Comment (gn5) [reply .github/scripts/install_singularity.sh:18]: just check that go compiler is installed.

File: .github/scripts/install_conda_envs.sh (hunk 6 above)
Comment (gn5) [reply .github/scripts/install_conda_envs.sh]: yes, I removed it.

File: .github/workflows/pull_request.yml (hunk 1 above)
Comment (sb10) [reply .github/workflows/pull_request.yml:24]: No, if it saves a manual install of java, the conda install is fine

File: .github/scripts/install_go.sh (hunk 7 above)
Comment (sb10) [reply .github/scripts/install_go.sh]: Wow, OK. Last time I looked, singularity was written in a different language. They re-wrote the whole thing in go!

File: .github/scripts/test_sanger_profile.sh [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+	|| exit 1
+
+# check that sanger profile activated params.allowNoprimer = false:
+find work -name .command.sh \
+    | xargs cat | grep 'ivar trim -e' \
+    && exit 1 \
+	|| echo "sanger profile: did NOT ran ivar trim -e" 
+
+# clean-up for following tests
+mv results results_sanger_profile
```
Comment (sb10) [inline .github/scripts/test_sanger_profile.sh]: Is this intended, or copy/paste error? Generally, do any of the tests look at the output of other tests? Does the results dir need to be kept for any of the tests? I think only my suggested way of testing `--ref` needs prior output, and that just needs specific output files.

**Review by sb10 — COMMENTED** [review]
Thanks, good changes.

File: .github/scripts/test_sanger_profile.sh (hunk 10 above)
Comment (gn5) [reply .github/scripts/test_sanger_profile.sh]: ok, I will remove the results instead of renaming them.

File: .github/scripts/test_PR_against_release.sh [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+cp -r results results_singularity_profile
+cp -r work work_singularity_profile
+
+# run tests against previous previous_release to compare outputs 
+git clone https://github.com/wtsi-team112/ncov2019-artic-nf.git previous_release 
+cd previous_release
+git checkout 42a79999f11304671b9d22e4959b0167b2130944 
+# the github runner only has 2 cpus available, so replace for that commit required:
+sed -i s'/cpus = 4/cpus = 2/'g conf/resources.config
+ln -s ../*.sif ./
```
Comment (sb10) [inline .github/scripts/test_PR_against_release.sh:27]: A previous version will likely have a different environment, so its sif would need re-creating. It's an important thing to test that new sif works as well as old sif. But maybe this can be ignored since the problem should go away when containers are in dockerhub.

File: .github/scripts/test_PR_against_release.sh [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+NXF_VER=20.03.0-edge nextflow run ./main.nf \
+       -profile singularity \
+       --directory $PWD/../.github/data/ \
+       --illumina \
+       --prefix test
+cp .nextflow.log ../artifacts/upstream.nextflow.log
```
Comment (sb10) [inline .github/scripts/test_PR_against_release.sh]: Call it previous_version.nextflow.log or similar?

**Review by sb10 — COMMENTED** [review]
Great, just cram tests to add now. Actually, it would be prudent to do a test with cram, `--bed --ref` and sanger profile combined, to ensure that multiple profiles work together along with multiple options, in the way we will actually run in production.

File: .github/scripts/test_PR_against_release.sh (hunk 12 above)
Comment (gn5) [reply .github/scripts/test_PR_against_release.sh]: ok, done.

File: .github/scripts/test_PR_against_release.sh (hunk 11 above)
Comment (gn5) [reply .github/scripts/test_PR_against_release.sh:27]: ok, yes at the moment it would also add 25 minutes to the tests.. Will ignore for now.

**Review by sb10 — APPROVED** [review]

**Review by sb10 — APPROVED** [review]
